### PR TITLE
Add custom Elytra support

### PR DESCRIPTION
**Pull request by l-Luna**
This adds support for items to act as custom Elytras via implementing IElytra on the item. 

I know that https://github.com/MinecraftForge/MinecraftForge/pull/4476 exists, but this makes creating Elytras with custom properties significantly easier.

One use-case is a RF-powered Elytra with can speed up at the cost of more energy.

### CODE REVIEW

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 1]
```diff
@@ -1,6 +1,14 @@
 --- ../src-base/minecraft/net/minecraft/entity/EntityLivingBase.java
 +++ ../src-work/minecraft/net/minecraft/entity/EntityLivingBase.java
-@@ -201,10 +201,11 @@
+@@ -74,6 +74,7 @@
+ import net.minecraft.world.WorldServer;
+ import net.minecraftforge.fml.relauncher.Side;
+ import net.minecraftforge.fml.relauncher.SideOnly;
++
```
Comment (ChloeDawn) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Imports should not be reorganized

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 2]
```diff
@@ -31,15 +39,24 @@
...
      }
  
      public EntityLivingBase func_110144_aD()
-@@ -670,8 +672,10 @@
+@@ -484,7 +487,7 @@
+             {
+                 soundevent = ((ItemArmor)item).func_82812_d().func_185017_b();
+             }
+-            else if (item == Items.field_185160_cR)
++            else if (item == Items.field_185160_cR || item instanceof net.minecraftforge.items.IElytra)
```
Comment (ChloeDawn) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Why does ItemElytra not implement this interface?

File: src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+		@Override
+		public void onHitWall(ItemStack stack, World world, BlockPos pos)
+		{
+			//I could put particle effects, explosions, fuel reductions, or really anything.
+		}
+	};
+	
+	
+	@Mod.EventBusSubscriber(modid = MODID)
```
Comment (ChloeDawn) [inline src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java]: There is absolutely no reason for having 2 separate event bus subscriber classes to register a test item + model. These methods can go in the main class.

File: jsons/1.12.2.json [hunk 4]
```diff
@@ -719,6 +719,6 @@
     "minecraftArguments": "--username ${auth_player_name} --version ${version_name} --gameDir ${game_directory} --assetsDir ${assets_root} --assetIndex ${assets_index_name} --uuid ${auth_uuid} --accessToken ${auth_access_token} --userType ${user_type} --versionType ${version_type}",
     "minimumLauncherVersion": 18,
     "releaseTime": "2017-09-18T08:39:46+00:00",
-    "time": "2018-02-15T16:26:45+00:00",
+    "time": "2018-04-02T07:39:26+00:00",
```
Comment (ChloeDawn) [inline jsons/1.12.2.json:722]: This entire file change is unrelated to your pull request and should be excluded

Comment (l-Luna) [reply jsons/1.12.2.json:722]: That was automatically done by gradlew genPatches

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch (hunk 2 above)
Comment (l-Luna) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: I wanted to keep patches as small as possible, so I just added checks for IElytra alongside regular Elytra checks.

File: src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java (hunk 3 above)
Comment (l-Luna) [reply src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java]: I was basing the test mod after the enchantment one

File: jsons/1.12.2.json (hunk 4 above)
Comment (ChloeDawn) [reply jsons/1.12.2.json:722]: That is not a valid reason for it to be included in your pull request..

Comment (l-Luna) [reply jsons/1.12.2.json:722]: But the build time, size, and checksum _were_ changed as a result of this PR, so it should change, right?

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 5]
```diff
@@ -1,6 +1,13 @@
 --- ../src-base/minecraft/net/minecraft/entity/EntityLivingBase.java
 +++ ../src-work/minecraft/net/minecraft/entity/EntityLivingBase.java
-@@ -201,10 +201,11 @@
+@@ -74,6 +74,7 @@
+ import net.minecraft.world.WorldServer;
+ import net.minecraftforge.fml.relauncher.Side;
+ import net.minecraftforge.fml.relauncher.SideOnly;
+ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
```
Comment (covers1624) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:8]: Don't touch imports in patch files.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 6]
```diff
@@ -247,15 +263,54 @@
...
 +        net.minecraftforge.common.ForgeHooks.onLivingJump(this);
      }
  
      protected void func_70629_bd()
-@@ -1874,7 +1901,8 @@
+@@ -1808,9 +1836,12 @@
+                 {
+                     if (this.func_184613_cA())
+                     {
++                    	ItemStack itemstack = this.func_184582_a(EntityEquipmentSlot.CHEST);
```
Comment (covers1624) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:278]: Weird spacing.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 7]
```diff
@@ -247,15 +263,54 @@
...
+                         if (this.field_70181_x > -0.5D)
+                         {
+-                            this.field_70143_R = 1.0F;
++                        	if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesSlowOnUpwardsFlight(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR){
++                        		this.field_70143_R = 1.0F;
++                        	}
```
Comment (covers1624) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Tabs, space after `if`, brace on newline.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 8]
```diff
@@ -247,15 +263,54 @@
...
+@@ -1847,6 +1878,12 @@
+                         this.field_70159_w *= 0.9900000095367432D;
+                         this.field_70181_x *= 0.9800000190734863D;
+                         this.field_70179_y *= 0.9900000095367432D;
++                        if(itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).getAccelerationMultiplier(itemstack, field_70170_p) != 1){
++                        	field_70159_w *= ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).getAccelerationMultiplier(itemstack, field_70170_p);
++                        	field_70181_x *= ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).getAccelerationMultiplier(itemstack, field_70170_p);
++                        	field_70179_y *= ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).getAccelerationMultiplier(itemstack, field_70170_p);
++                        }
++
```
Comment (covers1624) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Tabs, random newline, space after `if`, brace on newline.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 9]
```diff
@@ -247,15 +263,54 @@
...
+                         if (this.field_70123_F && !this.field_70170_p.field_72995_K)
+@@ -1858,7 +1895,10 @@
+                             if (f5 > 0.0F)
+                             {
+                                 this.func_184185_a(this.func_184588_d((int)f5), 1.0F, 1.0F);
+-                                this.func_70097_a(DamageSource.field_188406_j, f5);
++                                if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesDamageUponHittingWall(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR) {
++                                	this.func_70097_a(DamageSource.field_188406_j, f5);
++                                	if(itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra) ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).onHitWall(itemstack, field_70170_p, new BlockPos(field_70165_t, field_70163_u, field_70161_v));
++                                }
```
Comment (covers1624) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Tabs. Space after `if`and if body on new line.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 10]
```diff
@@ -293,7 +348,24 @@
...
+-            if (itemstack.func_77973_b() == Items.field_185160_cR && ItemElytra.func_185069_d(itemstack))
++            if (itemstack.func_77973_b() == Items.field_185160_cR && ItemElytra.func_185069_d(itemstack) || (itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).checkFlight(itemstack, field_70170_p)))
+             {
+                 flag = true;
+ 
+                 if (!this.field_70170_p.field_72995_K && (this.field_184629_bo + 1) % 20 == 0)
+                 {
++                	if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesDamageWhileFlying(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR){
++                		
++                	}
```
Comment (covers1624) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: This if block does nothing. Also tabs, space after `if` and brace on newline.

File: src/main/java/net/minecraftforge/items/IElytra.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
+package net.minecraftforge.items;
```
Comment (covers1624) [inline src/main/java/net/minecraftforge/items/IElytra.java:1]: License header.

File: src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
+package net.minecraftforge.debug.item;
```
Comment (covers1624) [inline src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java:1]: License Header.

File: src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    
+	public static final String MODID = "custom_elytra_item_test";
+	
+	public static class TestItem extends Item implements IElytra
+	{
+		public static Item testItem = new TestItem();
+		
+		@Override
+		public boolean doesSlowOnUpwardsFlight(ItemStack stack, World world)
+        {
```
Comment (covers1624) [inline src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java]: This class is full of tabs, except this line.

File: src/main/java/net/minecraftforge/items/IElytra.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+/**
+ * Implement this on your items to make them act as an Elytra
+ */
+public interface IElytra{
```
Comment (covers1624) [inline src/main/java/net/minecraftforge/items/IElytra.java]: Brace on newline.

File: jsons/1.12.2.json [hunk 15]
```diff
@@ -1,10 +1,10 @@
...
     "assetIndex": {
         "id": "1.12",
-        "sha1": "98c430c16a705f18a58a281b27caabf3ef09d40d",
-        "size": 169253,
-        "url": "https://launchermeta.mojang.com/mc/assets/1.12/98c430c16a705f18a58a281b27caabf3ef09d40d/1.12.json",
-        "totalSize": 127453671
+        "sha1": "e75e9535754c6f2158b0b18b35660f45c4495d78",
+        "size": 169257,
+        "url": "https://launchermeta.mojang.com/mc/assets/1.12/e75e9535754c6f2158b0b18b35660f45c4495d78/1.12.json",
+        "totalSize": 127722338
```
Comment (covers1624) [inline jsons/1.12.2.json:7]: Revert your local changes here, if you didn't do it, don't commit it.

**Review by covers1624 — COMMENTED** [review]
Formatting!
Whilst i see the intentions behind this, it feels extremely over complicated, Vanilla's elytra should implement the interface, and j8 defaults could be used in the interface for those optional things.

File: src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java (hunk 3 above)
Comment (covers1624) [reply src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java]: The second one is needed for client side events, it would crash on the server otherwise as the EventBus doesnt support SideOnly

File: jsons/1.12.2.json (hunk 4 above)
Comment (LexManos) [reply jsons/1.12.2.json:722]: No they went they were changed because mojang changed their assets for April fools. Don't include these jains in your prs

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 16]
```diff
@@ -247,15 +263,54 @@
...
-@@ -1874,7 +1901,8 @@
+@@ -1808,9 +1836,12 @@
+                 {
+                     if (this.func_184613_cA())
+                     {
++                    	ItemStack itemstack = this.func_184582_a(EntityEquipmentSlot.CHEST);
+                         if (this.field_70181_x > -0.5D)
+                         {
+-                            this.field_70143_R = 1.0F;
++                        	if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesSlowOnUpwardsFlight(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR){
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Can you just let the ielytra handle field 70143 R based on field 70181 x?

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 17]
```diff
@@ -293,7 +348,24 @@
...
+             ItemStack itemstack = this.func_184582_a(EntityEquipmentSlot.CHEST);
+ 
+-            if (itemstack.func_77973_b() == Items.field_185160_cR && ItemElytra.func_185069_d(itemstack))
++            if (itemstack.func_77973_b() == Items.field_185160_cR && ItemElytra.func_185069_d(itemstack) || (itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).checkFlight(itemstack, field_70170_p)))
+             {
+                 flag = true;
+ 
+                 if (!this.field_70170_p.field_72995_K && (this.field_184629_bo + 1) % 20 == 0)
+                 {
++                	if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesDamageWhileFlying(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR){
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: We shall let custom elytras determine their damage rate instead of just yes/no on damaging.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 18]
```diff
@@ -247,15 +263,54 @@
...
+ 
+                         if (this.field_70123_F && !this.field_70170_p.field_72995_K)
+@@ -1858,7 +1895,10 @@
+                             if (f5 > 0.0F)
+                             {
+                                 this.func_184185_a(this.func_184588_d((int)f5), 1.0F, 1.0F);
+-                                this.func_70097_a(DamageSource.field_188406_j, f5);
++                                if((itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).doesDamageUponHittingWall(itemstack, field_70170_p)) || itemstack.func_77973_b() == Items.field_185160_cR) {
++                                	this.func_70097_a(DamageSource.field_188406_j, f5);
++                                	if(itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra) ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).onHitWall(itemstack, field_70170_p, new BlockPos(field_70165_t, field_70163_u, field_70161_v));
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Can we just call hit wall without that boolean check? This ways modded elytras can do more things on hitting a wall.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch [hunk 19]
```diff
@@ -247,15 +263,54 @@
...
++                        		this.field_70143_R = 1.0F;
++                        	}
+                         }
+ 
+                         Vec3d vec3d = this.func_70040_Z();
+@@ -1847,6 +1878,12 @@
+                         this.field_70159_w *= 0.9900000095367432D;
+                         this.field_70181_x *= 0.9800000190734863D;
+                         this.field_70179_y *= 0.9900000095367432D;
++                        if(itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra && ((net.minecraftforge.items.IElytra) itemstack.func_77973_b()).getAccelerationMultiplier(itemstack, field_70170_p) != 1){
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: You should replace the 3 modifiders above instead of calling again. Also, I'd make a `void getAccelerationMultiplier(ItemStack, EntityLivingBase, int[] velocity)`, where the implementors just modify the int array to achieve acceleration multiplier.

**Review by liach — CHANGES_REQUESTED** [review]
Why are all these parameters `ItemStack` and `World`? As I see, you can change all `World` parameters to `EntityLivingBase` ones; you can access worlds from entity instances.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch (hunk 19 above)
Comment (pau101) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: I think a `Vec3d` return would be more appropriate.

Comment (liach) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Yes, it should be double. However I do not like vec3d for the excessive creation of unnecessary objects. (Vec3d is immutable)

Comment (pau101) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: An implementer doesn't need to produce a new instance each call but if it were an array you would have that instantiation (but keeping in mind JIT is also at play). However more importantly, this is the most ridiculous premature "optimization". Java is an object oriented language, use objects. This isn't the place to sacrifice API quality.

Comment (liach) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Then i suggest add 3 original numbers as arguments in case this elytra accelerates if one of x y z is too low. does that sound plausible?

**Comment by l-Luna** [comment]
I've reformatted lots of things, changed ItemElytra to implement IElytra, changed getAccelerationModifier to changeAcceleration, onHitWall now gets called regardless, Elytras now have control over damaging rates, all IElytra methods now accept Itemstack, EntityLivingBase.

**Comment by l-Luna** [comment]
jsons/1.12.2 has since been changed in the forge source, but the compare thing doesn't show it.

File: jsons/1.12.2.json (hunk 15 above)
Comment (l-Luna) [reply jsons/1.12.2.json:7]: These have since been changed by forge in source. The 'changes' thing only shows what it looked like as I forked it.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch (hunk 2 above)
Comment (l-Luna) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: This has now been changed

File: patches/minecraft/net/minecraft/client/entity/EntityPlayerSP.java.patch [hunk 20]
```diff
@@ -111,3 +111,12 @@
          boolean flag4 = (float)this.func_71024_bL().func_75116_a() > 6.0F || this.field_71075_bZ.field_75101_c;
  
          if (this.field_70122_E && !flag1 && !flag2 && this.field_71158_b.field_192832_b >= 0.8F && !this.func_70051_ag() && flag4 && !this.func_184587_cr() && !this.func_70644_a(MobEffects.field_76440_q))
+@@ -916,7 +941,7 @@
+         {
+             ItemStack itemstack = this.func_184582_a(EntityEquipmentSlot.CHEST);
+ 
+-            if (itemstack.func_77973_b() == Items.field_185160_cR && ItemElytra.func_185069_d(itemstack))
++            if ((itemstack.func_77973_b() == Items.field_185160_cR && ItemElytra.func_185069_d(itemstack)) || itemstack.func_77973_b() instanceof net.minecraftforge.items.IElytra)
```
Comment (liach) [inline patches/minecraft/net/minecraft/client/entity/EntityPlayerSP.java.patch:119]: IElytras should have ways to determine if the item stack can be used for flying (Like vanilla elytras can do that)

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch (hunk 5 above)
Comment (liach) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:8]: Well, this patch has nothing. Weird.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch (hunk 6 above)
Comment (liach) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:278]: Please change tabs to spaces.

File: patches/minecraft/net/minecraft/item/ItemElytra.java.patch [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
++public class ItemElytra extends Item implements net.minecraftforge.items.IElytra
+ {
+     public ItemElytra()
+     {
+@@ -61,4 +61,12 @@
+             return new ActionResult<ItemStack>(EnumActionResult.FAIL, itemstack);
+         }
+     }
++    // FORGE START
++    public boolean doesSlowOnUpwardsFlight(ItemStack stack, EntityLivingBase entity){return true;}
```
Comment (liach) [inline patches/minecraft/net/minecraft/item/ItemElytra.java.patch:17]: Add `@Override`s

File: src/main/java/net/minecraftforge/items/IElytra.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+
+    /**
+     * Called every tick to change the user's acceleration.
+     * 
+     * @param stack
+     *            The itemstack
+     * @param entity
+     *            The user
+     */
+    public void changeAcceleration(ItemStack stack, EntityLivingBase entity, double motionX, double motionY, double motionZ);
```
Comment (liach) [inline src/main/java/net/minecraftforge/items/IElytra.java:71]: The return type should be Vec3d

File: src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+            return true;
+        }
+
+        @Override
+        public void changeAcceleration(ItemStack stack, EntityLivingBase entity, double motionX, double motionY, double motionZ)
+        {
+            // Goes a bit faster than normal. This escalates QUICKLY; it can be changed as you use it, but I'd recommend sticking to 1 for most things.
+            motionX *= 1.1;
+            motionY *= 1.1;
+            motionZ *= 1.1;
```
Comment (liach) [inline src/test/java/net/minecraftforge/debug/item/CustomElytraItemTest.java:59]: Uhh this does not work.
